File: src/client/server/controller.rs
use std::collections::HashMap;

use sd_project::utils::models::task::DataTask;
use sd_project::utils::request::Request;
use sd_project::utils::response::Response;

pub struct TasksController {
    data: HashMap< String, HashMap<String, DataTask>>,
}

impl TasksController {
    pub fn new() -> Self {
        TasksController { 
            data: HashMap::new(),
        }
    }

    pub fn new_client(&mut self, cid: String) {
        self.data.insert(cid, HashMap::new());
    }

    pub fn create(&mut self, req: Request) -> Response {
        if None == req.cid || None == req.task || None == req.body {
            return Response::Error("Incomplet arguments".to_string());
        }
        let cid = req.cid.unwrap();
        let name = req.task.unwrap();
        let body = req.body.unwrap();
        
        if self.data.contains_key(&cid) {
            let tasks_table = self.data.get(&cid).unwrap();

            if !tasks_table.contains_key(&name) {
                match DataTask::from_string(body) {
                    Ok(task) => {
                        self.data.get_mut(&cid).unwrap().insert(name, task);
                        return Response::Sucess("Task created with sucess".to_string());
                    }
                    Err(e) => {
                        return Response::Error(e.to_string()); 
                    }
                }
            }
            return Response::Error("Already existing task".to_string());
        }
        return Response::Error("Client id not found".to_string());
    }

    pub fn update(&mut self, req: Request) -> Response {
        if None == req.cid || None == req.task || None == req.body {
            return Response::Error("Incomplet arguments".to_string());
        }
        let cid = req.cid.unwrap();
        let name = req.task.unwrap();
        let body = req.body.unwrap();
        
        if self.data.contains_key(&cid) {
            let tasks_table = self.data.get(&cid).unwrap();

            if tasks_table.contains_key(&name) {
                match DataTask::from_string(body) {
                    Ok(task) => {
                        self.data.get_mut(&cid).unwrap().insert(name, task);
                        return Response::Sucess("Task created with sucess".to_string());
                    }
                    Err(e) => {
                        return Response::Error(e.to_string()); 
                    }
                }
            }
            return Response::Error("Task not found".to_string());
        }
        return Response::Error("Client id not found".to_string());
    }

    pub fn list(&mut self, req: Request) -> Response {
        if None == req.cid {
            return Response::Error("Incomplet arguments".to_string())
        }
        let cid = req.cid.unwrap();

        if self.data.contains_key(&cid) {
            let tasks = self.data.get(&cid).unwrap().into_iter()
                .map(|(task, data)| { format!("{} - {:?}", task, data)})
                .reduce(|tasks, task| { format!("{}\n{}", tasks, task) });

            let tasks =  if tasks == None { "".to_string() } else { tasks.unwrap().to_string() };
            
            return Response::Sucess(tasks.to_string())
        }
        return Response::Error("Client id not found".to_string())
    }

    pub fn delete(&mut self, req: Request) -> Response {
        if None == req.cid || None == req.task {
            return Response::Error("Incomplet arguments".to_string());
        }
        let cid = req.cid.unwrap();
        let name = req.task.unwrap();
        
        if self.data.contains_key(&cid) {
            let tasks_table = self.data.get(&cid).unwrap();

            if tasks_table.contains_key(&name) {
                self.data.get_mut(&cid).unwrap().remove(&name);
                return Response::Sucess("Task deleted with sucess".to_string());
            }
            return Response::Error("Task not found".to_string())
        }
        return Response::Error("Client id not found".to_string())
    }
    
    pub fn delete_all(&mut self, req: Request) -> Response {
        if None == req.cid {
            return Response::Error("Incomplet arguments".to_string());
        }
        let cid = req.cid.unwrap();
        
        if self.data.contains_key(&cid) {
            self.data.get_mut(&cid).unwrap().clear();
            return Response::Sucess("Tasks deleted with sucess".to_string());
        }
        return Response::Error("Client id not found".to_string());
    }
}

impl Clone for TasksController {
    fn clone(&self) -> Self {
        Self { data: self.data.clone() }
    }
}

File: src/client/controller.py
import sys
sys.path.append('.')

from others.models import DataTask
from others.response import Response, ResponseType
from others.request import Request, RequestType

class TasksController():
    def __init__(self):
        self.data = {}

    def new_client(self, cid: int) -> Response:
        self.data[cid] = {}
        return Response(ResponseType.Sucess, f'Added Client {cid} with sucess')

    def delete_client(self, cid: int) -> Response:
        del self.data[cid]
        return Response(ResponseType.Sucess, f'Deleted client {cid} with sucess')

    def create(self, req: Request) -> Response:
        if req.req != RequestType.CreateTask:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.cid == None or req.task == None or req.body == None:
            return Response(ResponseType.Error, 'Incomplet argumets')
        
        try:
            data_task = DataTask.from_string(req.body)
        except Exception as e:
            return Response(ResponseType.Error, str(e))

        if req.cid in self.data:
            if not req.task in self.data[req.cid]:
                self.data[req.cid][req.task] = data_task
                return Response(ResponseType.Sucess, 'Task created with sucess')
            return Response(ResponseType.Error, 'Task already exist')
        return Response(ResponseType.Error, 'Client id not found')

    def update(self, req: Request) -> Response:
        if req.req != RequestType.UpdateTask:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.cid == None or req.task == None or req.body == None:
            return Response(ResponseType.Error, 'Incomplet argumets')
        
        try:
            data_task = DataTask.from_string(req.body)
        except Exception as e:
            return Response(ResponseType.Error, str(e))

        if req.cid in self.data:
            if req.task in self.data[req.cid]:
                self.data[req.cid][req.task] = data_task
                return Response(ResponseType.Sucess, 'Task created with sucess')
            return Response(ResponseType.Error, 'Task not found')
        return Response(ResponseType.Error, 'Client id not found')

    def list(self, req: Request) -> Response:
        if req.req != RequestType.ListTasks:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.cid == None:
            return Response(ResponseType.Error, 'Incomplet argumets')

        if req.cid in self.data:
            tasks: dict = self.data[req.cid]
            
            list = ''
            for name, data in tasks.items():
                list += f'\n{name} - {data.to_string()}'

            return Response(ResponseType.Sucess, list)
        return Response(ResponseType.Error, 'Client id not found')

    def delete(self, req: Request) -> Response:
        if req.req != RequestType.DeleteTask:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.cid == None or req.task == None:
            return Response(ResponseType.Error, 'Incomplet argumets')
        
        
        if req.cid in self.data:
            if req.task in self.data[req.cid]:
                del self.data[req.cid][req.task]
                return Response(ResponseType.Sucess, 'Task deleted with sucess')
            return Response(ResponseType.Error, 'Task not found')
        return Response(ResponseType.Error, 'Client not found')

    
    def delete_all(self, req: Request) -> Response:
        if req.req != RequestType.DeleteAllTasks:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.cid == None:
            return Response(ResponseType.Error, 'Incomplet argumets')
        
        
        if req.cid in self.data:
            self.data[req.cid].clear()
            return Response(ResponseType.Sucess, 'All Tasks deleted with sucess')
        return Response(ResponseType.Error, 'Client not found')
